### ISSUE TITLE
IgnoreClasses configuration does not support types from external assemblies

### ISSUE THREAD

**Issue by AndyMDoyle** [bugsnag/bugsnag-dotnet#106]
Specifying classes to be ignored from reporting does not always result in those classes being ignored. Classes contained under the System namespace appear okay, but others from external assemblies (such as ASP.NET MVC's HttpAntiForgeryException) never get ignored.

We have been trying to use the following:

```xml
<bugsnag apiKey="APIKEY"  ignoreClasses="System.Web.Mvc.HttpAntiForgeryException" />
```

The result of this should be that any HttpAntiForgery exceptions are ignored. However, the actual result is that these are being reported and appear in our dashboard.

From some initial tests, it appears this simplified reference to System.Web.Mvc.HttpAntiForgeryException is the problem, as the code in the Bugsnag.ConfigurationSection.Configuration.IgnoreClasses property getter isn't attempting to resolve types from all assemblies loaded in the current app domain.

I can suggest two possible fixes for this, however they come with their own pro's and con's.

**OPTION 1: Alter IgnoreClasses to allow specifying the assembly name with the type name**

Instead of ```System.Web.Mvc.HttpAntiForgeryException```, by specifying the assembly name in the form of ```System.Web.Mvc.HttpAntiForgeryException,System.Web.Mvc``` allows the call to ```Type.GetType(string)``` to correctly resolve the type.

Here's a basic working sample (although this does not address cases of specifying a single fully qualified type name without a trailing semicolon and will cause the type and assembly name to be split in this case):

```c#
var resolveWithAssemblyNames = InternalIgnoreClasses.Contains(";");

_ignoreClasses = InternalIgnoreClasses
	.Split(resolveWithAssemblyNames ? ';' : ',')
	.Select(c => Type.GetType(c))
	.Where(t => t != null)
	.ToArray();
```

Pro's

- Types will be resolved correctly
- Very simple code change

Con's

- This would change how this value is expected to be set in an apps webconfig.xml or app.xml file as the string could no longer be split on the comma. Instead, to support this, multiple types would have to be separated by a different character in order to allow adding the assembly name to the type name.

**OPTION 2: Attempt to resolve the types from all assemblies loaded by the current app domain**

To correctly find types without knowing the assembly name, the only reliable way I have found is to scan through all assemblies that are currently loaded.

This can be achieved with the following:

```c#
var matchClasses = InternalIgnoreClasses
		.Split(',');
		
_ignoreClasses = AppDomain.CurrentDomain.GetAssemblies()
		.Where(assembly => !assembly.IsDynamic)
		.SelectMany(assembly => assembly.GetTypes())
		.Where(type => matchClasses.Contains(type.FullName))
		.ToArray();
```

Pro's

- This will find the type regardless of what assembly it is contained in (as long as the application references that assembly

Con's

- All assemblies need to be scanned
- Depending on how many assemblies are referenced, and how many types those assemblies contain, this method can cause significant delay. In a simple test with ASP.NET MVC and 5 other random NuGet packages added to a test project, resolving System.Web.Mvc.HttpAntiForgeryException added around 1 second in a very basic benchmark (debug mode was on, so not an entirely accurate comparison as you'd see in release mode).

**Summary**

I'm normally against changing/breaking something public and potentially widely used such as the expected format for ```ignoreClasses``` in the configuration file; but in this case, I believe it is the better of the two choices given the performance impact of scanning all assemblies.

Thoughts?

**Comment by AndyMDoyle** [comment]
Having thought about this for a while, I think a better solution is to retain the existing support for separating multiple types with a comma (so nothing changes for existing users), but to introduce a special character that can be used to append the assembly name to a type name.

For example, if someone is using ```System.InvalidOperationException,System.ArgumentException``` this should remain unchanged. However if I want to now add System.Web.Mvc.HttpAntiForgeryException, I could include it in this configuration string by adding ```System.Web.Mvc.HttpAntiForgeryException@System.Web.Mvc```. The ```@``` character could be used in a way that read <this-class>@<this-assembly>, and the IgnoreClasses function could replace the ```@``` character before attempting to resolve the type.

So overall the above example could be changed to ```System.InvalidOperationException,System.ArgumentException,System.Web.Mvc.HttpAntiForgeryException@System.Web.Mvc``` to keep backwards compatibility/support yet allow referencing non-core exception types.

**Comment by martin308** [comment]
@AndyMDoyle Thanks for putting so much thought into this! I like your thinking with the `@` symbol. I did a bit of research and unfortunately the `@` symbol can be used as a valid identifier :rage1: 

Thinking about this though it might be better to change how these classes are provided in the `config` files. We could define them similar to how we define [metadata](https://docs.bugsnag.com/platforms/dotnet/mvc/configuration-options/#global-metadata) to attach to reports, for example:

```
<bugsnag apiKey="your-api-key">
  <ignoreClasses>
    <class name="System.Web.Mvc.HttpAntiForgeryException,System.Web.Mvc" />
  </ignoreClasses>
</bugsnag>
```

We might be able to support backwards compatibility as well by supporting both definition types. Although I would have to look into if this is possible. It might be that the extended version would require a different name than `ignoreClasses`

```
<bugsnag apiKey="your-api-key" ignoreClasses="System.ArgumentException">
  <ignoreClasses>
    <class name="System.Web.Mvc.HttpAntiForgeryException,System.Web.Mvc" />
  </ignoreClasses>
</bugsnag>
```

What do you think?

**Comment by martin308** [comment]
@AndyMDoyle I whipped up a quick implementation of it yesterday, you can catch a sneak preview [here](https://github.com/bugsnag/bugsnag-dotnet/compare/martin308/ignore-classes). Just working on the docs and the naming of the element in the config as it is too much work to support reusing the `ignoreClasses` name for the element alongside using it as an attribute

**Comment by martin308** [comment]
This is released now in v2.2.0!

**Comment by AndyMDoyle** [comment]
@martin308, thanks, that is great news.

I will get this setup today. Thanks again!
